trantor/utils/LogStream.h

```diff
@@ -21,7 +21,9 @@
 #include <assert.h>
 #include <string.h>  // memcpy
 #include <string>
-
+#if CXX_STD >= 17
```

Comment:
Only adding things to the API depending on the something is fine, but it should not be something like a C++ version.  Rather it should just be a normal define, not in a config file or anything, especially as these are header-inlined, that way the user can opt to compile it in or out on their choice.  What if they are using a third-party string_view, such as boost's or one of the others?  Would be good to have the define's based on that where the user can state it as they need, no config file definitions needed.  In addition this will fail to compile if the library is compiled with C++17 but the user library is compiled with C++14 (very real possibility), which is why such define's should not be baked into a config header.

Reply:
How about changing the line to `#ifdef USE_STD_STRING_VIEW`(If users want to enable the API, they should use cmake like `cmake .. -DUSE_STD_STRING_VIEW`)?

If the user library is compiled with C++14, they must use `cmake .. -DCMAKE_CXX_STANDARD=14` to compile the trantor.

I seem to understand some of your meaning, the trantor must now be distributed in the source code (compiled and installed by the user), if I modify it according to your principles, trantor seems to be available in the form of a library in addition to the source distribution (Header and library files), is my understanding correct?

Reply:
Can I use the __cplusplus macro?

Reply:
> How about changing the line to #ifdef USE_STD_STRING_VIEW(If users want to enable the API, they should use cmake like cmake .. -DUSE_STD_STRING_VIEW)?

That would be far more common yeah.  :-)

> If the user library is compiled with C++14, they must use cmake .. -DCMAKE_CXX_STANDARD=14 to compile the trantor.

If you don't want to mandate a minimum C++ version that is fine, but if your library actually does require C++14 features then it is best to mandate that as a minimum regardless.

> I seem to understand some of your meaning, the trantor must now be distributed in the source code (compiled and installed by the user), if I modify it according to your principles, trantor seems to be available in the form of a library in addition to the source distribution (Header and library files), is my understanding correct?

In a fully and properly setup CMake build system, it would be possible for all of these to be done by the user:

* Grab a copy of the repo, compile and install it (which installs a perfect TrantorConfig.cmake system configuration file so no FindTrantor.cmake is needed, then they can use their globally installed version.
* Grab a copy of the repo, the user `add_subdirectory`'s it into their program and they can compile and use it all internally.
* You can distribute an 'installation' set of the library, which includes a set of libraries, header files, a FindTrantor.cmake file, etc...).
* The above installation set can be distributed as a tar.gz/zip or other compressed format, a deb file, a windows installer file, an rpm file, a whole variety of other things (all natively supported as install generators in CMake).

In general the first 2 points are generally sufficient, but once those are set up properly then adding in an installable distributable is fairly simple.  :-)

> Can I use the __cplusplus macro?

The `__cplusplus` macro used to be... not standard, but as of C++14 it is now a standard macro (C++14 is defined in it as `201402L` for example, C++17 is `201703L`).  Because it is a yearmonthqualifier it's not always fun to use though, so always double-check the values you check in it!  If you mandate a C++14 minimum version then it is always safe to use.  If you want to work on older C++'s then it 'might' work (on most compilers it will, though a variety of MSVC's are broken) but always check specific values so it at least falls back cleanly on compilers that don't support it.  I personally would use it to do conditional compilation based on standards version.

Reply:
@OvermindDL1 I can't agree more. I learned a lot. Thank you so much.
I mandate a C++14 minimum version, because I move some objects into lambdas(this is supported since C++14) in trantor.